File: nvpl_blas/c/sspr2.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_sspr2
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 2;
    nvpl_int_t incX = 1;
    nvpl_int_t incY = 2;
    float alpha = 1.0f;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_UPLO uplo = CblasUpper;
    float * AP = NULL;
    float * X = NULL;
    float * Y = NULL;
    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);
    nvpl_int_t len_y = 1 + (N - 1) * labs(incY);
    nvpl_int_t len_ap = (N * (N + 1)) / 2;

    printf("\nExample: cblas_sspr2 for a rank-2 update of a symmetric packed matrix\n\n");
    printf("#### args: n=%" PRId64 ", incX=%" PRId64 ", incY=%" PRId64 ", order=%c, uplo=%c\n",
           (int64_t)N, (int64_t)incX, (int64_t)incY, order_to_char(order), uplo_to_char(uplo));
    printf("           alpha=%g\n", alpha);

    // allocate memory
    AP = (float *)malloc(len_ap  * sizeof(float));
    X = (float *)malloc(len_x * sizeof(float));
    Y = (float *)malloc(len_y * sizeof(float));

    // fill data
    fill_svector(AP, len_ap, 1);
    fill_svector(X, N, incX);
    fill_svector(Y, N, incY);

    // print input data
    print_svector(AP, len_ap, 1, "AP");
    print_svector(X, N, incX, "X");
    print_svector(Y, N, incY, "Y");

    // call cblas_sspr2
    cblas_sspr2(order, uplo, N, alpha, X, incX, Y, incY, AP);

    // print result
    printf("\nAfter sspr2 operation:\n");
    print_svector(AP, len_ap, 1, "AP");

    // release memory
    free(AP);
    free(X);
    free(Y);
    return 0;
}

File: nvpl_blas/f77/sspr2.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     SSPR2 
!
!******************************************************************************
      program   SSPR2_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      character*1     uplo
      parameter       (uplo='U')
      integer         m, n, incx, incy, i
      parameter       (m=2, n=2, incx=1, incy=1)
      real            alpha
      parameter       (alpha=1.0)
      integer         xsize, ysize, apsize
      parameter       (xsize=1+(n-1)*abs(incx),
     &                 ysize=1+(n-1)*abs(incy),
     &                 apsize=(n*(n+1))/2)
      real            x(xsize), y(ysize), ap(apsize)

*       External Subroutines
      external        print_svector, fill_svector
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, incx, incy, uplo
      print 101, alpha
      call fill_svector(x,n,incx)
      call fill_svector(y,n,incy)
      call fill_svector(ap,apsize,1)
      call print_svector(ap,apsize,1,'AP')
      call print_svector(x,n,incx,'X')
      call print_svector(y,n,incy,'Y')
*
*      Call SSPR2 subroutine
      call SSPR2(uplo,n,alpha,x,incx,y,incy,ap)
*
      print*
      print 102
      call print_svector(ap,apsize,1,'AP')

      stop
 99   format('Example: SSPR2 for a rank-2 update of a'
     &       ' symmetric packed matrix')
 100  format('#### args: n=',i1,', incx=',i1,', incy=',i1,
     &       ', uplo=',a1)
 101  format(11x,'alpha=',f4.2)
 102  format('The result of SSPR2: ')
      end
